src/Compiler/Driver/ReuseTcResults/TcResultsPickle.fs

```diff
@@ -0,0 +1,1810 @@
+﻿module internal FSharp.Compiler.ReuseTcResults.TcResultsPickle
+
+open FSharp.Compiler
+open FSharp.Compiler.AbstractIL.IL
+open FSharp.Compiler.AccessibilityLogic
+open FSharp.Compiler.CheckBasics
+open FSharp.Compiler.CheckDeclarations
+open FSharp.Compiler.CompilerGlobalState
+open FSharp.Compiler.ConstraintSolver
+open FSharp.Compiler.DiagnosticsLogger
+open FSharp.Compiler.NameResolution
+open FSharp.Compiler.ParseAndCheckInputs
+open FSharp.Compiler.Infos
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.TcGlobals
+open FSharp.Compiler.Text.Range
+open FSharp.Compiler.TypedTree
+open FSharp.Compiler.TypedTreeBasics
+open FSharp.Compiler.TypedTreeOps
+open FSharp.Compiler.TypedTreePickle
+open FSharp.Compiler.Xml
+
+open Internal.Utilities
+open Internal.Utilities.Collections
+open Internal.Utilities.Library
+open Internal.Utilities.Library.Extras
+
+type TcSharedData = { TopAttribs: TopAttribs }
+
+// pickling
+
+let p_stamp = p_int64
+
+let p_stamp_map pv = p_Map p_stamp pv
+
+let p_non_null_slot f (x: 'a | null) st =
+    match x with
+    | null -> p_byte 0 st
+    | h ->
+        p_byte 1 st
+        f h st
+
+let p_ILTypeDefAdditionalFlags (x: ILTypeDefAdditionalFlags) st = p_int32 (int x) st
+
+let p_ILTypeDef (x: ILTypeDef) st =
+    p_string x.Name st
+    //p_type_attributes x.Attributes
+    //p_il_type_def_layout x.Layout
+    //x.Implements
+    //x.Extends
+    //x.Methods
+    //x.NestedTypes
+    //x.Fields
+    //x.MethodImpls
+    //x.Events
+    //x.Properties
+    p_ILTypeDefAdditionalFlags x.Flags st
+//x.SecurityDeclsStored
+//x.CustomAttrsStored
+//p_il
+//p_int32 x.MetadataIndex st
+
+let p_tyar_spec_data_new (x: Typar) st =
+    p_tup6
+        p_ident
+        p_attribs
+        p_int64
+        p_tyar_constraints
+        p_xmldoc
+        p_stamp
+        (x.typar_id, x.Attribs, int64 x.typar_flags.PickledBits, x.Constraints, x.XmlDoc, x.Stamp)
+        st
+
+let p_tyar_spec_new (x: Typar) st =
+    //Disabled, workaround for bug 2721: if x.Rigidity <> TyparRigidity.Rigid then warning(Error(sprintf "p_tyar_spec: typar#%d is not rigid" x.Stamp, x.Range))
+    if x.IsFromError then
+        warning (Error((0, "p_tyar_spec: from error"), x.Range))
+
+    p_osgn_decl st.otypars p_tyar_spec_data_new x st
+
+let p_tyar_specs_new = (p_list p_tyar_spec_new)
+
+let rec p_ty_new (ty: TType) st : unit =
```

Comment:
I introduced here this `_new` suffix system. Helps to not get lost here, but naming should be better...